MaAslin2-Tutorial (#5173)
galaxyproject/training-material: topics/microbiome/tutorials/multivariable-association/tutorial.md

---
layout: tutorial_hands_on

title: "Determining multivariable association between various meta’omic features using MaAslin2"
subtopic: metagenomics 
tags:
    - Heatmap
    - metagenomics 
    - associations

level: Introductory
zenodo_link: https://zenodo.org/records/12614561
questions:
- How do I find associations between microbial features and specific metadata variables ? 
objectives:
- Identify statistically significant associations between microbial features and metadata variables (such as clinical conditions, environmental factors, or demographic information) in microbiome data.
- Uncover potential biomarkers associated with specific disease states.

> time_estimation: 15 minutes

Reviewer: ```suggestion
time_estimation: 15M
```

(format here should be 1H35M)

> <comment-title></comment-title>
>
> For more information about MaAslin2, [click here](https://huttenhower.sph.harvard.edu/maaslin/).
{: .comment}

MaAsLin2 requires the following input files:

- **Taxonomy (or features) file** : \
        This file is tab-delimited.\
        Formatted with features as columns and samples as rows.\
        The transpose of this format is also okay.\
        Possible features in this file include microbes, genes, pathways, etc.
- **Metadata file** : \
        This file is tab-delimited.\
        Formatted with features as columns and samples as rows.\
        The transpose of this format is also okay.

The Taxonomy file can contain samples not included in the metadata file (or vice versa). For both cases, those samples not included in both files will be removed from the analysis. Also the samples do not need to be in the same order in the two files.

> <agenda-title></agenda-title>
>
> In this tutorial, we will cover:
>
> 1. TOC
> {:toc}
>
{: .agenda}

> # Get the data

Author: I couldn't find studies which used both  HMP2 data as well as Maaslin2 tool. So instead I mentioned studies which used Maaslin2 tool with the parameters used. If it does not feel right , then please let me know.

> # Get the data

Reviewer 2: Maybe its enought to add that this was the initial data used to demonstrate maaslin2 ?

> # Get the data

Author: sounds better

>
>    As a default, Galaxy uses the link as the name of the new dataset. It also does not link the dataset to a database or a reference genome.
>
>    {% snippet faqs/galaxy/datasets_rename.md %}
>
> 4. Inspect the content of a file.
>
>    > <tip-title>Inspecting the content of a dataset</tip-title>
>    >
>    > * Click on the {% icon galaxy-eye %} (eye) icon next to the relevant history entry
>    > * View the content of the file in the central panel
>    {: .tip}
>
>    > <question-title></question-title>
>    >
>    > 1.  What is the main difference between the two files?
>    >
>    >
>    > > <solution-title></solution-title>
>    > > 1. The metadata file describes sample characteristics (e.g., clinical data, demographics) while the taxonomy or feature file contains microbial data (e.g., taxa abundance) used for analysis in microbiome studies.
>    > {: .solution }
>    {: .question}
>
{: .hands_on}

# Find associations between the two files
Now we will find significant associations between microbial features( taxonomy file) and metadata variables (metadata file) using the **MaAslin2** tool

> <hands-on-title> Task description </hands-on-title>
>

> > 1. {% tool [MaAsLin 2](toolshed.g2.bx.psu.edu/repos/iuc/maaslin2/maaslin2/1.16.0+galaxy0) %} with the following parameters:

Reviewer 2: can you rerun the tool later and show differences in models and parameters ?

   > <comment-title></comment-title>
   >
   > - Reference for a variable with more than two levels is provided as a string of `variable,reference`.
   > - Reference for more than one variable having more than two levels each is provided as a string of `variable1,reference1,variable2,reference2` .
   > - Example, both diagnosis and site variable have more than two levels hence reference can be provided as `diagnosis,CD,site,Cedars-Sinai`.
   {: .comment}

4. **Additional options** :
- min_abundance : The minimum abundance for each feature [ Default: 0 ]
- min_prevalence:  The minimum percent of samples for which a feature is detected at minimum abundance [ Default: 0.1 ]
- max_significance: The q-value threshold for significance [ Default: 0.25 ]
- normalization: The normalization method to apply [ Default: "TSS" ] [ Choices: "TSS", "CLR", "CSS", "NONE", "TMM" ]
- transform: The transform to apply [ Default: "LOG" ] [ Choices: "LOG", "LOGIT", "AST", "NONE" ]

> - analysis_method : The analysis method to apply [ Default: "LM" ] [ Choices: "LM", "CPLM", "ZICP", "NEGBIN", "ZINB" ]

Reviewer 2: can we get a summary of the methods, maybe with literature ?

- correction : The correction method for computing the q-value [ Default: "BH" ]
- standardize : Apply z-score so continuous metadata are on the same scale [ Default: TRUE ]
- plot_heatmap : Generate a heatmap for the significant associations [ Default: TRUE ]
- heatmap_first_n : In heatmap, plot top N features with significant associations [ Default: 50 ]
- plot_scatter : Generate scatter plots for the significant associations [ Default: TRUE ]
- cores : The number of R processes to run in parallel [ Default: 1 ]

# Reading Output Files
The tool generate the following five major files:
- **Data output files**
    1. `residuals.rds`

>             This file contains a data frame with residuals for each feature.

Reviewer 2: What are the residuals ? Maybe we could add something like this at the end of the tutorial, or at least explain how to use the residuals could be used: https://forum.biobakery.org/t/maaslin2-model-validation/1237/4

    2. `significant_results.tsv`
            Provides the most important output from MaAsLin2 which is the list of significant associations.
    3.  `all_results.tsv`
            Same format as significant_results.tsv, but include all association results (instead of just the significant ones).

- **Visualization output files**
    4.  `heatmap.pdf`
        This file contains a heatmap of the significant associations.
        ![heatmap](../../images/heatmap_maaslin2.png "heatmap of significatn associations")

>     5.  `plots :`

Reviewer 2: Can you also show one or two examples of the plots and explain them

    >       >
    >       > For example, if you look for `Bifidobacterium longum` in the heatmap, you'll notice that its occurrence in the human gut is least affected by the individual's age and shows a neutral effect in relation to their diagnosis of UC (Ulcerative Colitis) and non-IBD (non-Inflammatory Bowel Disease).
    > > 2. The significant.tsv file shows statistically significant associations between microbial features and metadata variables that meet a specified threshold (in our case, the default `Maximum significance = 0.25`). It includes effect sizes, p-values, and adjusted p-values (q-values) to indicate the strength, direction, and reliability of each association. This file helps identify meaningful relationships in the microbiome data.
    > {: .solution }
    {: .question}

# Conclusion

> In essence, uncovering associations between microbial features and metadata variables through tools like MaAsLin2 not only deepens our understanding of microbiome dynamics but also holds promise for clinical applications, personalized health strategies, and advancing the field of microbiome research.

Reviewer 2: Maybe explain how these findings can be further visualized using tools like ampvis2 and phyloseq ...

- Options:\
      1. <u>Benjamini & Hochberg(BH)(aka false discovery rate(fdr))</u>: A common method used for FDR correction. It ranks the p-values from smallest to largest and adjusts them based on their rank and the total number of tests.
      2. <u>Benjamini & Yekutieli(BY)</u>: Similar to Benjamini-Hochberg but includes a correction factor that accounts for the correlation between tests.
      3. <u> Bonferroni correction</u>: Divides the significance threshold (alpha level) by the number of tests performed and then compares each p-value to this adjusted significance level to determine if it is statistically significant. 
      4. <u>Hochberg </u>:  It is similar to the Bonferroni correction but is often more powerful, meaning it has better statistical power to detect true effects while controlling for false positives.
      5. <u>Hommel</u>: controls the Family-Wise Error Rate (FWER) by adjusting p-values in a step-down fashion, starting from the smallest p-value and progressively increasing the threshold. It is more powerful than Bonferroni while maintaining strict error control.

>       6. <u>Holm</u>:  controls the Family-Wise Error Rate (FWER) by sequentially adjusting p-values from smallest to largest, comparing ch step. This stepwise approach is less conservative than the Bonferroni correction, offering greater statistical power.\

Reviewer 3: What is this ` ch step`?

Author: its the adjusted critical value at step i,  calculated as α/(m-i+1) ​. Sorry for the confusion, I'll explain it more clearly in the tutorial.

[PMID: 1009442](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009442 )

![sensitivity and false discovery rate (FDR) across different tools](https://journals.plos.org/ploscompbiol/article/figure/image?size=large&id=10.1371/journal.pcbi.1009442.g004 "Source: <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009442#pcbi-1009442-g004">sensitivity and false discovery rate (FDR) across different tools</a>"){:width="60%"}

- The above figure compares various tools for differential abundance detection (Panel A) and multivariable association detection (Panel B) in microbiome studies, based on sensitivity and false discovery rate (FDR).
- **Sensitivity** measures how well the methods detect true signals, higher values lead to better performance.
- **False discovery rate (FDR)** measures the proportion of false positives among detected signals (lower FDR is better).
- MaAsLin2 is the clear standout for both differential abundance detection and multivariable association detection, showing high sensitivity and maintaining a low FDR.

> 

Author: https://github.com/galaxyproject/training-material/pull/5173/commits/5bd7914b731e212f346298c89a8bd6735748f1de

3. **Random effects**: In some studies, like those following people over time or studying families, samples from the same group can be similar. MaAsLin2 helps handle this by letting researchers choose a grouping factor. This helps make sure the statistical analysis is more accurate. For example, setting random_effects = "Subject_ID" helps control for the correlation between samples that come from the same individual.

4. **Reference**: It allows researchers to establish a baseline or standard category against which other categories are compared, helping to interpret and understand the effects of different variables on microbial features. 

   > <comment-title></comment-title>

>    > - In MaAslin2, the reference level is must for variables with more than two distinct kinds of values.

Reviewer 3: Something is wrong with this sentence.

Author: ![Screenshot from 2024-09-29 15-17-03](https://github.com/user-attachments/assets/a1fbaac2-1203-45f0-8a4a-bc02975f75a1)

@bgruening I was attempting to rephrase the clarification provided in this. Did you find any factual error in this or you want me to rephrase it more clearly? 
https://forum.biobakery.org/t/passing-multiple-reference-factors-in-maaslin2-version-1-16-0/7035/5

Reviewer 3: "is must for" sounds just wrong to me.

Author: @bgruening , how about this -"In MaAslin2, the reference level is required for variables with more than two levels" . 

User will get an error if he choose a fixed effect with more than two level and forgets to choose a reference among those levels. so this is really importmant

Reviewer 3: Yes way better.

---
layout: tutorial_hands_on

> title: "Determining multivariable association between various meta’omic features using MaAslin2"

Reviewer 2: ```suggestion
title: "Determining multivariable association of various meta’omic features using MaAslin2"
```
`Between` would mean that we determine the associations between feature (but we do between meta'omics and the metadata)'

Author: Thanks for pointing it out :)

| **ANCOM-BC**        | Compositionality and bias correction          | Computationally intensive for large datasets | ANCOM-BC is good for simpler designs, but MaAsLin2 handles complex metadata better. |
| **LEfSe**           | Easy to interpret, focuses on effect size     | No covariates, may overfit               | LEfSe is simpler but lacks the flexibility and multivariable depth of MaAsLin2. |
| **ALDEx2**          | Robust to sparsity, small sample sizes        | Limited handling of complex metadata     | ALDEx2 is suitable for small datasets, but MaAsLin2 is superior in handling multivariable data and covariates. |
| **MetagenomeSeq**   | Handles zero-inflation, sparse data           | Computationally heavy for large datasets | MetagenomeSeq is great for zero-inflated data, but lacks MaAsLin2's multivariable modeling capacity. |
| **Corncob**         | Models both abundance and variability        | Complex to use, requires R expertise     | Corncob excels at overdispersion analysis, but MaAsLin2 is easier for broader multivariable models. |

> | **Phyloseq + DESeq2**| Strong for RNA-seq and transcriptomics; integrates with Phyloseq | Lacks compositionality awareness         | While DESeq2 works for microbiome data, MaAsLin2 offers more suitable options for compositional data and covariate handling. |

Reviewer 2: ```suggestion
| **DESeq2**| Originally developed for RNA-seq and transcriptomics | Lacks compositionality awareness         | While DESeq2 works for microbiome data, MaAsLin2 offers more suitable options for compositional data and covariate handling. |
```
In theory any tool can work with phyloseq, would leave it out.

Author: committed https://github.com/galaxyproject/training-material/pull/5173/commits/30d373a029cd1b7d1b4b14311e42897bb7e9bda2

| **Limma-Voom**      | Effective for RNA-seq and microarray data, handles low counts | Not tailored for compositional microbiome data | Limma-Voom is well-suited for gene expression, but MaAsLin2 better accounts for the unique characteristics of microbiome data. |

> - ANCOM-BC and MaAsLin2, outperform general-purpose tools like DESeq2 and limma-voom when it comes to microbiome data. This is due to their handling of the compositional nature of microbiome data and the sparsity typical of microbial datasets.[PMID: 36617187](https://pubmed.ncbi.nlm.nih.gov/36617187/)

Reviewer 2: Can you please exchange all links to studies with citations as shown here: https://github.com/galaxyproject/training-material/blob/5012353c3c3462ce860e77e8fb34393bb8bbd8fb/topics/microbiome/tutorials/pathogen-detection-from-nanopore-foodborne-data/tutorial.md?plain=1#L89

![sensitivity and false discovery rate (FDR) across different tools](https://journals.plos.org/ploscompbiol/article/figure/image?size=large&id=10.1371/journal.pcbi.1009442.g004 "Source: <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009442#pcbi-1009442-g004">sensitivity and false discovery rate (FDR) across different tools</a>"){:width="60%"}

- The above figure compares various tools for differential abundance detection (Panel A) and multivariable association detection (Panel B) in microbiome studies, based on sensitivity and false discovery rate (FDR).
- **Sensitivity** measures how well the methods detect true signals, higher values lead to better performance.
- **False discovery rate (FDR)** measures the proportion of false positives among detected signals (lower FDR is better).
- MaAsLin2 is the clear standout for both differential abundance detection and multivariable association detection, showing high sensitivity and maintaining a low FDR.

> <comment-title></comment-title>
>
> For more information on MaAslin2, [click here](https://huttenhower.sph.harvard.edu/maaslin/).
{: .comment}

MaAsLin2 requires the following input files:

> - **Taxonomy (or features) file**: \

Reviewer 2: ```suggestion
- **Features file**: \
```

Would like to keep it more generic. Maybe add some examples like: (OTU/ASV abundance table,  MAGs abundance matrix, taxonomy table, gene count matrix at the end of the list

        This file is tab-delimited.\
        Formatted with features as columns and samples as rows.\

>         The transposition of this format is also okay.\

Reviewer 2: Is it ? Can our wrapper work with both ?

Author: Thanks for pointing this out @paulzierep .  I cross checked and unfortunately our wrapper cannot work with both. I have made the changes accordingly in the tutorial. Sorry for the confusion.

        Possible features in this file include microbes, genes, pathways, etc.
- **Metadata file** : \
        This file is tab-delimited.\
        Formatted with features as columns and samples as rows.\
        The transposition of this format is also okay.

> The Taxonomy file can contain samples not included in the metadata file (or vice versa). For both cases, those samples not included in both files will be removed from the analysis.

Reviewer 2: ```suggestion
The feature file can contain samples not included in the metadata file (or vice versa). For both cases, those samples not included in both files will be removed from the analysis.
```

- ANCOM-BC and MaAsLin2, outperform general-purpose tools like DESeq2 and limma-voom when it comes to microbiome data. This is due to their handling of the compositional nature of microbiome data and the sparsity typical of microbial datasets. ({% cite Yang2023 %})
- While general methods like DESeq2 and limma-voom are reliable for gene expression analysis, they do not handle the unique properties of microbiome data as effectively as MaAsLin2. The latter provides a more accurate estimation of differential abundance in the presence of metadata confounders.

![sensitivity and false discovery rate (FDR) across different tools](https://journals.plos.org/ploscompbiol/article/figure/image?size=large&id=10.1371/journal.pcbi.1009442.g004 "Source: <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009442#pcbi-1009442-g004">sensitivity and false discovery rate (FDR) across different tools</a>"){:width="60%"} 

> - The above figure compares various tools for differential abundance detection (Panel A) and multivariable association detection (Panel B) in microbiome studies, based on sensitivity and false discovery rate (FDR).

Reviewer 2: This and the next 4 sentences do no need to be bullet points.

>    > {: .solution }
>    {: .question}
>
{: .hands_on}

MaAsLin2 is designed to analyze various types of **tabular data** for multivariable association studies. The following table formats are compatible:

**OTU/ASV Abundance Tables:** These tables represent the abundance of Operational Taxonomic Units (OTUs) or Amplicon Sequence Variants (ASVs) across different samples, allowing for analysis of microbial diversity in relation to metadata.

**MAGs Abundance Matrices:** These matrices reflect the abundance of Metagenome-Assembled Genomes (MAGs) in samples, facilitating the exploration of associations between reconstructed genomes and various metadata.

**Taxonomy Tables:** These tables contain taxonomic classifications of microbes, enabling researchers to investigate how different taxonomic levels relate to associated metadata variables.

**Gene Count Matrices:** These matrices provide counts of specific genes across samples, useful for examining associations between gene abundances and metadata, particularly in metagenomics or transcriptomics studies.

> 

Author: https://github.com/galaxyproject/training-material/pull/5173/commits/76283dc7b27c40bdb20e57f4ff4a86b4d28e1753

**Outputs compatible with MaAsLin2**:
    - Shared File (OTU Table): Mothur generates a "shared" file that contains the abundance of OTUs across samples, which can be reformatted and used as input to MaAsLin2.
    - Features File: Mothur also outputs taxonomy assignments for OTUs, which can be paired with the shared file.
    - Metadata File: As in QIIME2, metadata is used during the Mothur workflow and can be reused in MaAsLin2.
3. [**MetaPhlAn (Metagenomic Phylogenetic Analysis)**](https://training.galaxyproject.org/training-material/by-tool/iuc/metaphlan.html) is a tool that provides taxonomic profiling of microbial communities from metagenomic sequencing data. It is available in Galaxy and can generate outputs that are directly compatible with MaAsLin2.\
**Outputs compatible with MaAsLin2**:
    - Taxonomic Profiles: MetaPhlAn generates taxonomic abundance tables (e.g., relative abundance of microbial taxa at various taxonomic levels) that can be directly used as the feature table for MaAsLin2.
    - Metadata File: As with other tools, the sample metadata used in MetaPhlAn analysis can also be reused in MaAsLin2.

> # Find associations between the two files

Reviewer 2: ```suggestion
# Find associations between the two groups
```

Author: https://github.com/galaxyproject/training-material/pull/5173/commits/22fc038f89586809913aa45415cf4b13ea45d07b

Now we will find significant associations between microbial features(features file) and metadata variables (metadata file) using the **MaAslin2** tool

> <hands-on-title> Task description </hands-on-title>
>
> 1. {% tool [MaAsLin 2](toolshed.g2.bx.psu.edu/repos/iuc/maaslin2/maaslin2/1.16.0+galaxy0) %} with the following parameters:
>    - *"Interactions: Fixed effects"*: `c3:age`, `c4:diagnosis`
>    - *"Random effects"*: `c5:subject`
>    - *"Reference"*: `diagnosis,CD`
>
>  Keep the rest of the default values as it is.
{: .hands_on}

> # Understanding parameters in the tool

Reviewer 2: ```suggestion
## Understanding parameters in the tool
```

Author: https://github.com/galaxyproject/training-material/pull/5173/commits/bcc9c781027a19e2e189aa447986b443ea687dfd

The logit transformation is only applicable to data within the open interval (0, 1), so values exactly at 0 or 1 need to be adjusted                     (e.g., adding a small constant like 0.001).
        3.  <u>Arcsine Square Root Transformation (AST)</u>: It is a statistical transformation used primarily on proportion or percentage data.\
The transformation starts by taking the square root of the proportion value. This step reduces the impact of extreme values.\
Next, it applies the arcsine function (the inverse of the sine function) to the square root result. The arcsine function helps to normalize the distribution further.\
Used when you are working with proportion data, such as relative abundances in microbiome studies, where the values are bounded between 0 and 1. 
                                 
9. **analysis_method** [ Default: "LM" ] [ Options: "LM", "CPLM", "ZICP", "NEGBIN", "ZINB" ]
- The analysis method to apply.
- Options: \

>         1. <u>Linear Model (LM)</u>: Determines how changes in metadata are associated with changes in the taxonomy data. 

Reviewer 2: ```suggestion
        1. <u>Linear Model (LM)</u>: The defaul linear model of MaAsLin2. By default, MaAsLin2 uses linear regression (for continuous metadata) or logistic regression (for binary metadata) as its association method.
```
All models should do that, but what is the difference between them

**FWER** is the probability of finding at least one false positive among all the tests performed, assuming all null hypotheses are true.\
**FWER Control** is used to minimize the risk of incorrectly claiming significant results when there are none, thus maintaining the overall reliability of the results.\
For more information on correction methods, [click here](https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/p.adjust).

11.  **standardize**: Apply z-score so continuous metadata are on the same scale [ Default: TRUE ]
12.  **plot_heatmap**: Generate a heatmap for the significant associations [ Default: TRUE ]
13. **heatmap_first_n**: In heatmap, plot top N features with significant associations [ Default: 50 ]
14.  **plot_scatter**: Generate scatter plots for the significant associations [ Default: TRUE ]
15.   **cores**: The number of R processes to run in parallel [ Default: 1 ]

> # Model types in MaAslin2

Reviewer 2: Not clear to me what this paragraph means, you already explain the models **analysis_method** - so its not clear what point 2 and 3 mean, can you choose these options ? Would remove or rephrase.

Author: These are not the option that can be chosen. These are broader categories of models under which the analysis methods explained above fall . For instance CPLM, ZICP etc come under generalized linear model. If user chooses random effect , then model works as linear mixed model.

Author: @paulzierep  I can make it more clear in the tutorial if you want. Let me know if the above explaination makes sense to you. I can add this explaination in the analysis method section  and remove the following para on these models.

Reviewer 2: then you should say, that those are models chosen by maaslin depending on the data input and the parameter choice. Otherwise, it's confusing. However, if CPLM, ZICP lead to Generalized Linear Models then this should also be stated. You got the *When to Use:* part - but users cannot choose anything - so that would confuse me

Author: Understood. I  will make the necessary changes. Thanks for the detailed explanation :)

The analysis, which utilized additive boosting of generalized linear models for feature reduction, revealed significant associations between ARGs and E. coli presence, as well as early-life factors such
as antibiotic use and other exposures. Key parameters included CLR normalization of compositional abundance data, no standardization of continuous variables, and a strict significance threshold
(q-value < 0.01) using Benjamini-Hochberg correction. This approach highlighted how early exposures influence the resistome and its relationship with E. coli in the infant's gut.

# Conclusion

In essence, uncovering associations between microbial features and metadata variables through tools like MaAsLin2 not only deepens our understanding of microbiome dynamics but also holds promise for clinical applications, personalized health strategies, and advancing the field of microbiome research.

The results obtained from MaAslin2 can further be visualized using tools like [**phyloseq**](https://training.galaxyproject.org/training-material/by-tool/interactive_tool_phyloseq.html).\

> Tools like phyloseq require that you prepare your data (OTU/ASV table, metadata) in a structured format before visualizing.\

Reviewer 2: I would add here, that the features need to be subsampled to only include significant features. This can be done with https://usegalaxy.eu/?tool_id=toolshed.g2.bx.psu.edu%2Frepos%2Fiuc%2Fampvis2_subset_taxa%2Fampvis2_subset_taxa%2F2.8.9%2Bgalaxy0&version=latest
